data/transactions/logic/box.go

```diff
@@ -0,0 +1,315 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package logic
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/transactions"
+)
+
+const (
+	boxCreate = iota
+	boxRead
+	boxWrite
+	boxDelete
+)
+
+func (cx *EvalContext) availableBox(name string, operation int, createSize uint64) (string, bool, error) {
+	if cx.txn.Txn.OnCompletion == transactions.ClearStateOC {
+		return "", false, fmt.Errorf("boxes may not be accessed from ClearState program")
+	}
+
+	dirty, ok := cx.available.boxes[boxRef{cx.appID, name}]
+	if !ok {
+		return "", false, fmt.Errorf("invalid Box reference %v", name)
+	}
+
+	// Since the box is in cx.available, we know this GetBox call is cheap. It
+	// will go (at most) to the cowRoundBase. Knowledge about existence
+	// simplifies write budget tracking, then we return the info to avoid yet
+	// another call to GetBox which most ops need anyway.
+	content, exists, err := cx.Ledger.GetBox(cx.appID, name)
+	if err != nil {
+		return "", false, err
+	}
+
+	switch operation {
+	case boxCreate:
+		if exists {
+			if createSize != uint64(len(content)) {
+				return "", false, fmt.Errorf("box size mismatch %d %d", uint64(len(content)), createSize)
+			}
+			// Since it exists, we have no dirty work to do. The weird case of
+			// box_put, which seems like a combination of create and write, is
+			// properly handled because already used boxWrite to declare the
+			// intent to write (and tracky dirtiness).
+			return content, exists, nil
+		}
+		fallthrough // If it doesn't exist, a create is like write
+	case boxWrite:
+		writeSize := createSize
+		if exists {
+			writeSize = uint64(len(content))
+		}
+		if !dirty {
+			cx.available.dirtyBytes += writeSize
+		}
+		dirty = true
+	case boxDelete:
```

Comment:
as I understand ioBudget is designed to prevent DB/disk overuse so that two questions:
1. box creation leads to DB write of to the app account record and to KV. Since it is empty array, it is msgpacked to couple bytes.
2. box deletion is the same, but it removes a row from a DB.
My point is that ops might be as expensive as writes since DB loads and append to WAL the entire 4k page anyway. Having `MaxBoxSize = 32K`, create/delete should be 8 times cheaper than write op, and do not cost 0.

Reply:
Resolving - Based on live discussion we agreed that the as-is behavior is acceptable because the only boxes that can be touched must be specified in the transaction + creates/deletes in a single transaction are consolidated before reaching the database.